From 56783eb7b97598838da086efc6a5f4da4347c1d1 Mon Sep 17 00:00:00 2001
From: Nan Li <nli@suse.com>
Date: Mon, 13 Apr 2015 08:59:45 +0200
Subject: [PATCH] xfs: xfs_alloc_fix_minleft can underflow near ENOSPC
References: bnc#913080, bnc#912741
Patch-mainline: v3.20

Message from original patch follows:

Test generic/224 is failing with a corruption being detected on one
of Michael's test boxes.  Debug that Michael added is indicating
that the minleft trimming is resulting in an underflow:

.....
 before fixup:              rlen          1  args->len          0
 after xfs_alloc_fix_len  : rlen          1  args->len          1
 before goto out_nominleft: rlen          1  args->len          0
 before fixup:              rlen          1  args->len          0
 after xfs_alloc_fix_len  : rlen          1  args->len          1
 after fixup:               rlen          1  args->len          1
 before fixup:              rlen          1  args->len          0
 after xfs_alloc_fix_len  : rlen          1  args->len          1
 after fixup:               rlen 4294967295  args->len 4294967295
 XFS: Assertion failed: fs_is_ok, file: fs/xfs/libxfs/xfs_alloc.c, line: 1424

The "goto out_nominleft:" indicates that we are getting close to
ENOSPC in the AG, and a couple of allocations later we underflow
and the corruption check fires in xfs_alloc_ag_vextent_size().

The issue is that the extent length fixups comaprisons are done
with variables of xfs_extlen_t types. These are unsigned so an
underflow looks like a really big value and hence is not detected
as being smaller than the minimum length allowed for the extent.
Hence the corruption check fires as it is noticing that the returned
length is longer than the original extent length passed in.

This can be easily fixed by ensuring we do the underflow test on
signed values, the same way xfs_alloc_fix_len() prevents underflow.
So we realise in future that these casts prevent underflows from
going undetected, add comments to the code indicating this.

Reported-by: Michael L. Semon <mlsemon35@gmail.com>
Tested-by: Michael L. Semon <mlsemon35@gmail.com>
Signed-off-by: Dave Chinner <dchinner@redhat.com>

Signed-off-by: Mark Fasheh <mfasheh@suse.de>

diff --git a/fs/xfs/xfs_alloc.c b/fs/xfs/xfs_alloc.c
index e60a4f1..cbcceee 100644
--- a/fs/xfs/xfs_alloc.c
+++ b/fs/xfs/xfs_alloc.c
@@ -256,9 +256,11 @@ xfs_alloc_fix_len(
 	if (k == args->mod)
 		return;
 	if (k > args->mod) {
+		/* casts to (int) catch length underflows */
 		if ((int)(rlen = rlen - k - args->mod) < (int)args->minlen)
 			return;
 	} else {
+		/* casts to (int) catch length underflows */
 		if ((int)(rlen = rlen - args->prod - (args->mod - k)) <
 		    (int)args->minlen)
 			return;
@@ -287,7 +289,8 @@ xfs_alloc_fix_minleft(
 	if (diff >= 0)
 		return 1;
 	args->len += diff;		/* shrink the allocated space */
-	if (args->len >= args->minlen)
+	/* casts to (int) catch length underflows */
+	if ((int)args->len >= (int)args->minlen)
 		return 1;
 	args->agbno = NULLAGBLOCK;
 	return 0;
-- 
1.7.12.4

